src/client/Microsoft.Identity.Client/Cache/CacheFallbackOperations.cs

```diff
@@ -231,105 +231,79 @@ internal static class CacheFallbackOperations
                 {
-                    List<KeyValuePair<AdalTokenCacheKey, AdalResultWrapper>> upnEntries =
-                        listToProcess.Where(p => upn.Equals(p.Key.DisplayableId, StringComparison.OrdinalIgnoreCase)).ToList();
-
-                    listToProcess = upnEntries;
+                    listToProcess =
+                        listToProcess.Where(p => account.Username.Equals(
+                            p.Key.DisplayableId, StringComparison.OrdinalIgnoreCase));                        
+                    
+                    filtered = true;
                 }
 
                 // if uniqueId is provided then use it to filter
-                if (!string.IsNullOrEmpty(uniqueId))
+                if (!string.IsNullOrEmpty(account?.HomeAccountId?.ObjectId))
                 {
-                    List<KeyValuePair<AdalTokenCacheKey, AdalResultWrapper>> uniqueIdEntries =
-                        listToProcess.Where(p => uniqueId.Equals(p.Key.UniqueId, StringComparison.OrdinalIgnoreCase)).ToList();
+                    listToProcess =
+                        listToProcess.Where(p => account.HomeAccountId.ObjectId.Equals(
+                            p.Key.UniqueId, StringComparison.OrdinalIgnoreCase)).ToList();
 
-                    listToProcess = uniqueIdEntries;
+                    filtered = true;
                 }
 
-                List<MsalRefreshTokenCacheItem> list = new List<MsalRefreshTokenCacheItem>();
-                foreach (KeyValuePair<AdalTokenCacheKey, AdalResultWrapper> pair in listToProcess)
+                // We should filter at leasts by one criteria to ensure we adequate RT
+                if (!filtered)
                 {
-                    string homeAccountId = null;
-                    if (!string.IsNullOrEmpty(pair.Value.RawClientInfo))
-                    {
-                        homeAccountId = ClientInfo.CreateFromJson(pair.Value.RawClientInfo).ToAccountIdentifier();
-                    }
-
-                    list.Add(new MsalRefreshTokenCacheItem(
-                      new Uri(pair.Key.Authority).Host,
-                      pair.Key.ClientId,
-                      pair.Value.RefreshToken,
-                      pair.Value.RawClientInfo,
-                      familyId: null,
-                      homeAccountId: homeAccountId));
+                    logger.Warning("Could not filter ADAL entries by either UPN or unique ID, skipping.");
+                    return null;
                 }
 
-                return list;
+                return listToProcess.Select(adalEntry => new MsalRefreshTokenCacheItem(
+                      new Uri(adalEntry.Key.Authority).Host,
+                      adalEntry.Key.ClientId,
+                      adalEntry.Value.RefreshToken,
+                      adalEntry.Value.RawClientInfo,
+                      familyId: null,
+                      homeAccountId: GetHomeAccountId(adalEntry.Value)))
+                .FirstOrDefault();
```

Comment:
Second part of fix: it is acceptable to find multiple RTs for the same user in the ADAL cache, because ADAL stores all tokens (AT, RT, IdT) on a per-tenant basis, even though RTs are tenantless.